### PR TITLE
Feature/calendar

### PR DESCRIPTION
**Pull request by frozenhelium**
### This PR ensures that there are
- [x] No unnecessary console logs
- [x] No unwanted comments
- [x] No typos
- [x] `yarn typecheck` shows no issue
- [x] `yarn lint` shows no issue

### This PR adds one or more new components / utils

- [x] Add the newly created component and its props to exports
- [x] Write story for newly created component(s)
- [x] No empty files

### This PR fixes or updates existing components / utils

- [x] Update / add story to reflect the changes
- [x] Doesn't introduce breaking changes

### Beta release vXX.XX.XX-beta.YY
- [ ] Add a commit for beta release

### TODO
- [x] Add predefined date ranges in Date Range Input

### CODE REVIEW

**Review by samshara — APPROVED** [review]

**Comment by AdityaKhatri** [comment]
To be fixed:
- [x] Month cannot be selected
- [x] Predefined date ranges should have different colors on hover
- [x] Bottom align year and month selection in popup
- [x] Selected date is not shown in date select
- [x] Date input should go to selected month if value is already there

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]
Fix few items

**Comment by frozenhelium** [comment]
@AdityaKhatri done!

File: src/components/Calendar/CalendarDate/index.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import RawButton, { Props as RawButtonProps } from '../../RawButton';
+import { ymdToDateString } from '../../../utils';
+
+import styles from './styles.css';
+
+export interface Props {
+    className?: string;
+    year: number;
+    month: number;
+    date: number;
```
Comment (tnagorra) [inline src/components/Calendar/CalendarDate/index.tsx:13]: Shouldn't this be day?
```suggestion
    day: number;
```

File: src/components/Calendar/index.tsx [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    { key: 10, label: 'November' },
+    { key: 11, label: 'December' },
+];
+
+function getStartOfWeek(year: number, month: number) {
+    return new Date(year, month, 1).getDay();
+}
+
+function getNumDaysInMonth(year: number, month: number) {
+    return new Date(year, month + 1, 0).getDate();
```
Comment (tnagorra) [inline src/components/Calendar/index.tsx:58]: Let's add a comment why we did this.

File: src/components/Calendar/index.tsx [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+function getDates(year: number, month: number) {
```
Comment (tnagorra) [inline src/components/Calendar/index.tsx:66]: Let's return a list of object:
```
// Example
{ day: number, month: 'prev' | 'current' | 'next' }[]
```

File: src/components/Calendar/index.tsx [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    );
+}
+
+const monthKeySelector = (m: MonthName) => m.key;
+const monthLabelSelector = (m: MonthName) => m.label;
+
+type RendererOmissions = 'year' | 'month' | 'date' | 'currentYear' | 'currentMonth' | 'currentDate' | 'onClick' | 'activeDate';
+export interface Props<P extends CalendarDateProps> {
+    className?: string;
+    dateRenderer?: (p: P) => React.ReactElement;
```
Comment (tnagorra) [inline src/components/Calendar/index.tsx]:
```suggestion
    dateRenderer?: (prop: P) => React.ReactElement;
```

File: src/components/DateRangeInput/index.tsx [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+    { key: 'yesterday', label: 'Yesterday' },
+    { key: 'thisWeek', label: 'This week' },
+    { key: 'lastSevenDays', label: 'Last 7 days' },
+    { key: 'thisMonth', label: 'This month' },
+    { key: 'lastThirtyDays', label: 'Last 30 days' },
+    { key: 'lastThreeMonths', label: 'Last 3 months' },
+    { key: 'lastSixMonths', label: 'Last 6 months' },
+    { key: 'thisYear', label: 'This year' },
+    { key: 'lastYear', label: 'Last year' },
+];
```
Comment (tnagorra) [inline src/components/DateRangeInput/index.tsx]: Let's also add value property that computes the required value

File: src/components/DateRangeInput/index.tsx [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+                            startDate: prevTempStartDate,
+                            endDate: ymdToDateString(year, month, day),
+                        },
+                        name,
+                    );
+                }
+
+                hideCalendar();
+
+                return undefined;
```
Comment (tnagorra) [inline src/components/DateRangeInput/index.tsx]: Do we need to return undefined?

**Review by subinasr — APPROVED** [review]

File: src/components/Calendar/CalendarDate/index.tsx (hunk 1 above)
Comment (frozenhelium) [reply src/components/Calendar/CalendarDate/index.tsx:13]: day represents day of week in js

File: src/components/DateRangeInput/index.tsx (hunk 6 above)
Comment (frozenhelium) [reply src/components/DateRangeInput/index.tsx]: Apparently

Comment (tnagorra) [reply src/components/DateRangeInput/index.tsx]: oh no!

File: src/components/Calendar/index.tsx [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+    { key: 10, label: 'November' },
+    { key: 11, label: 'December' },
+];
+
+function getStartOfWeek(year: number, month: number) {
+    return new Date(year, month, 1).getDay();
+}
+
+function getNumDaysInMonth(year: number, month: number) {
+    // Setting date to 0 will switch the date to last day of previous month
```
Comment (tnagorra) [inline src/components/Calendar/index.tsx:57]: We should have an alternative name to differentiate between "गते" and "मिति"

File: src/components/Calendar/index.tsx [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+        monthSelectionPopupClassName,
+        initialDate,
+        activeDate,
+    } = props;
+
+    const current = initialDate ? new Date(initialDate) : new Date();
+    const currentYear = current.getFullYear();
+    const currentMonth = current.getMonth();
+
+    const today = new Date();
```
Comment (tnagorra) [inline src/components/Calendar/index.tsx]: We should move today at the start 
and use it's value on current as fallback. Just for sematics.

File: src/components/Calendar/index.tsx [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+
+                            children = (
+                                <DateRenderer
+                                    {...combinedProps}
+                                />
+                            );
+                        }
+
+                        return (
+                            <div className={styles.dayContainer}>
```
Comment (tnagorra) [inline src/components/Calendar/index.tsx]: We need a key here!

File: src/components/DateRangeInput/index.tsx [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+    useBlurEffect(
+        showCalendar,
+        handlePopupBlur,
+        popupRef,
+        containerRef,
+    );
+
+    const dateRendererParams = React.useCallback(() => ({
+        startDate: tempStartDate ?? value?.startDate,
+        endDate: !tempStartDate ? value?.endDate : undefined,
```
Comment (tnagorra) [inline src/components/DateRangeInput/index.tsx]: Let's add a comment here!

File: src/components/DateRangeInput/index.tsx [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+    }), [tempStartDate, value]);
+
+    const handleCalendarDateClick: CalendarProps<never>['onDateClick'] = React.useCallback(
```
Comment (tnagorra) [inline src/components/DateRangeInput/index.tsx:178]: This function is complicated. It does a lot of things.
Let's try to break this down.

File: src/components/DateRangeInput/index.tsx [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+        },
+        [name, onChange, hideCalendar],
+    );
+
+    const handlePredefinedOptionClick = React.useCallback((optionKey: PredefinedDateRangeKey) => {
+        if (onChange) {
+            const option = predefinedDateRangeOptions.find((d) => d.key === optionKey);
+            if (!option) {
+                return;
+            }
```
Comment (tnagorra) [inline src/components/DateRangeInput/index.tsx]: Maybe this return is not placed correctly.

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/components/DateRangeInput/index.tsx [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+                    const endDate = prev > current ? prevTempStartDate : lastDate;
+
+                    onChange({
+                        startDate,
+                        endDate,
+                    }, name);
+                }
+
+                return undefined;
+            });
```
Comment (tnagorra) [inline src/components/DateRangeInput/index.tsx]: I still don't like calling sideEffects inside setState. I don't  know if it's the correct thing to do.

Comment (tnagorra) [reply src/components/DateRangeInput/index.tsx]: @AdityaKhatri any thoughts?

Comment (tnagorra) [reply src/components/DateRangeInput/index.tsx]: Maybe try using useReducer and move the side-effects there? Idk if we should call sideEffects there as well

Comment (tnagorra) [reply src/components/DateRangeInput/index.tsx]: Looks like reducer function should also be pure

**Review by tnagorra — APPROVED** [review]

**Review by AdityaKhatri — APPROVED** [review]

**Pull request merged**